### PR TITLE
[UnifiedPDF] Fix context menu page navigation for 2up.

### PR DESCRIPTION
**Pull request by sammygill**
#### 55b05170db3b61e4d6266a0a9580f4665adcbb39
<pre>
[UnifiedPDF] Fix context menu page navigation for 2up.
<a href="https://bugs.webkit.org/show_bug.cgi?id=269501">https://bugs.webkit.org/show_bug.cgi?id=269501</a>
<a href="https://rdar.apple.com/problem/123032781">rdar://problem/123032781</a>

Reviewed by Simon Fraser.

274737@main provided the initial implementation for navigating between
pages using context menu options. It worked fine for pages in single
page display modes but not all the time in two up display modes. This
patch tries to address that by making sure we navigate between pairs
of pages in two up mode.

When determining which page to scroll to in two up mode we need to
consider whether the &quot;current page,&quot; is the first page in the row or the
second page in the row. If we are navigating forwards and the &quot;current
page,&quot; is the first page in the row, then we need to navigate two pages
forward to go to the next row. Similarly, if we are navigating backwards
and the &quot;current page,&quot; is the second page in the row, we need to
go two pages backwards.

I also changed indexForCurrentPageInView to use a new method,
nearestPageIndexForDocumentPoint, instead of pageIndexForDocumentPoint
since it was not guaranteed the later would alawys return a page index.
Fow single page mode, this new method iterates over the pages in order
and compares the point with the bottom of each page. Once we find a page
point&apos;s Y location that is lower than the document points&apos; then we have
found the page the point is contained in. Two up modes works in the
same way except we also need to compare the X location for the points.

I also renamed the DisplayModes enums to explicitly state whether they
are discrete or continuous to help introduce some new helpers without
them having confusing names (isSinglePageDisplayMode and
isTwoUpDisplayMode).

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.h:
(WebKit::PDFDocumentLayout::isSinglePageDisplayMode const):
(WebKit::PDFDocumentLayout::isTwoUpDisplayMode const):
(WebKit::PDFDocumentLayout::pagesPerRow const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.mm:
(WebKit::PDFDocumentLayout::isLeftPageIndex const):
(WebKit::PDFDocumentLayout::isRightPageIndex const):
(WebKit::PDFDocumentLayout::isLastPageIndex const):
(WebKit::PDFDocumentLayout::nearestPageIndexForDocumentPoint const):
(WebKit::PDFDocumentLayout::updateLayout):
(WebKit::PDFDocumentLayout::layoutPages):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::shouldUseScrollSnapping const):
(WebKit::UnifiedPDFPlugin::shouldDisplayPage):
(WebKit::UnifiedPDFPlugin::indexForCurrentPageInView const):
(WebKit::UnifiedPDFPlugin::contextMenuItemTagFromDisplayMode const):
(WebKit::UnifiedPDFPlugin::displayModeFromContextMenuItemTag const):
(WebKit::UnifiedPDFPlugin::navigationContextMenuItems const):
(WebKit::UnifiedPDFPlugin::performContextMenuAction):

Canonical link: <a href="https://commits.webkit.org/274858@main">https://commits.webkit.org/274858@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/a2ea08c1bd9d8069a6b1a23f10cdf84fc7ffc25b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40099 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19111 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42476 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42644 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/36188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22024 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16440 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33359 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40673 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16091 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34637 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13915 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13952 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43922 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36379 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35921 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39681 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14914 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12239 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37953 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16533 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9021 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16582 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16177 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm [hunk 1]
```diff
@@ -1379,10 +1379,51 @@ static String mutationObserverNotificationString()
...
+    switch (m_documentLayout.displayMode()) {
+    case PDFDocumentLayout::DisplayMode::TwoUpDiscrete:
+    case PDFDocumentLayout::DisplayMode::TwoUpContinuous:
+        for (PDFDocumentLayout::PageIndex index = 0; index < pageCount; ++index) {
+            if (index == pageCount - 1)
+                return index;
+
+            auto currentPageBounds = m_documentLayout.layoutBoundsForPageAtIndex(index);
+            if (documentSpacePoint.y() < currentPageBounds.maxY()) {
+                auto pairedPageIndex =  [index, this]() {
```
Comment (smfr) [inline Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm]: Extra space

File: Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm [hunk 2]
```diff
@@ -1379,10 +1379,51 @@ static String mutationObserverNotificationString()
...
+                    if (index % m_documentLayout.pagesPerRow())
```
Comment (smfr) [inline Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm]: Maybe PDFDocumentLayout should have `isLeftpage()`, `isRightPage()`

File: Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm [hunk 3]
```diff
@@ -1379,10 +1379,51 @@ static String mutationObserverNotificationString()
     return std::nullopt;
 }
 
-std::optional<PDFDocumentLayout::PageIndex> UnifiedPDFPlugin::indexForCurrentPageInView() const
+
+PDFDocumentLayout::PageIndex UnifiedPDFPlugin::nearestPageIndexForDocumentPoint(WebCore::IntPoint documentSpacePoint) const
```
Comment (smfr) [inline Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm]: This code could live in `PDFDocumentLayout`

**Review by smfr — APPROVED** [review]

File: Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h [hunk 4]
```diff
@@ -365,7 +365,8 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
     WebCore::IntPoint offsetContentsSpacePointByPageMargins(WebCore::IntPoint pointInContentsSpace) const;
 
     std::optional<PDFDocumentLayout::PageIndex> pageIndexForDocumentPoint(const WebCore::IntPoint&) const;
-    std::optional<PDFDocumentLayout::PageIndex> indexForCurrentPageInView() const;
+    PDFDocumentLayout::PageIndex nearestPageIndexForDocumentPoint(WebCore::IntPoint) const;
```
Comment (aprotyas) [inline Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h]: Shouldn't we just adopt this everywhere (most places?) where we call `pageIndexForDocumentPoint`?

Comment (aprotyas) [reply Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h]: Well, that doesn't have to happen now, I'll look at existing callers after this merges.

File: Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm (hunk 2 above)
Comment (aprotyas) [inline Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm]: Nit, just capture the layout
```suggestion
                auto pairedPageIndex =  [index, &documentLayout = m_documentLayout]() {
                    if (index % documentLayout.pagesPerRow())
```

File: Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm [hunk 5]
```diff
@@ -1379,10 +1379,51 @@ static String mutationObserverNotificationString()
...
+            }
+        }
+        break;
+    case PDFDocumentLayout::DisplayMode::SinglePageDiscrete:
+    case PDFDocumentLayout::DisplayMode::SinglePageContinuous: {
+        for (PDFDocumentLayout::PageIndex index = 0; index < pageCount; ++index) {
+            auto pageBounds = m_documentLayout.layoutBoundsForPageAtIndex(index);
+
+            if (documentSpacePoint.y() <= pageBounds.maxY() || index == pageCount - 1)
+                return index;
```
Comment (aprotyas) [inline Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm]: Do (don't?) we need to keep rotation in mind?

File: Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm [hunk 6]
```diff
@@ -2019,14 +2060,27 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
...
         break;
-    case ContextMenuItemTag::NextPage:
-        if (auto currentPageIndex = indexForCurrentPageInView(); currentPageIndex < m_documentLayout.pageCount() - 1)
-            scrollToPage(currentPageIndex.value() + 1);
+    case ContextMenuItemTag::NextPage: {
+        auto currentPageIndex = indexForCurrentPageInView();
+        auto pageCount = m_documentLayout.pageCount();
+        auto pagesPerRow = m_documentLayout.pagesPerRow();
+
+        if (currentPageIndex < pageCount - 1 && (m_documentLayout.isSinglePageDisplayMode() || (m_documentLayout.isTwoUpDisplayMode() && currentPageIndex % pagesPerRow)))
```
Comment (aprotyas) [inline Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm]: Was a bit hard to follow this condition, maybe if we could self-document the condition with a function name and then call it? (I don't have a suggestion since I'm not 100% sure what the condition is)

Comment (aprotyas) [reply Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm]:
```cpp
if (canMoveToNextPage(...)) {
    pageIncrement = ...
    scrollToPage(currentPageIndex + pageIncrement);
}
```

File: Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm [hunk 7]
```diff
@@ -2019,14 +2060,27 @@ static bool isContextMenuEvent(const WebMouseEvent& event)
...
-            return scrollToPage(currentPageIndex.value() - 1);
+    }
+    case ContextMenuItemTag::PreviousPage: {
+        auto currentPageIndex = indexForCurrentPageInView();
+        auto pagesPerRow = m_documentLayout.pagesPerRow();
+
+        if (currentPageIndex && (m_documentLayout.isSinglePageDisplayMode() || !(currentPageIndex % pagesPerRow)))
+            scrollToPage(currentPageIndex - 1);
+        else if (currentPageIndex > 1)
+            scrollToPage(currentPageIndex - pagesPerRow);
```
Comment (aprotyas) [inline Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm]: Ditto

**Review by aprotyas — COMMENTED** [review]
LGTM with mostly non-blocking comments!

File: Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm (hunk 2 above)
Comment (hortont424) [reply Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm]: I wrote code similar to this in UnifiedPDFPlugin::shouldDisplayPage, we should factor it out at some point.

File: Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm (hunk 6 above)
Comment (sammygill) [reply Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm]: Yeah agreed that this is a lot to parse. Added a lambda and some helpers so hopefully this is a bit better
